neps/nep-0591.md

```diff
@@ -0,0 +1,207 @@
+
+## Specification
+
+Global contract can be deployed in 2 ways: either by its hash or by owner account id.
+Contracts deployed by hash are effectively immutable and cannot be updated.
+When deployed by account id the owner can redeploy the contract updating it for all its users.
+Users can use contracts deployed by hash if they prefer having control over contract updates. In order to update the contract user would have to explicitly switch to a different version of the contract deployed under a different hash.
+Contracts deployed by account id should be used when user trusts contract developers to update the contract for them. For example if user accounts are created specifically for some application to be onchain.
+
+We introduce new receipt action for deploying global contracts:
+
+```rust
+struct DeployGlobalContractAction {
+    code: Vec<u8>,
+    deploy_mode: GlobalContractDeployMode,
+}
+
+enum GlobalContractDeployMode {
+    /// Contract is deployed under its code hash.
+    /// Users will be able reference it by that hash.
+    /// This effectively makes the contract immutable.
+    CodeHash,
+    /// Contract is deployed under the owner account id.
+    /// Users will be able reference it by that account id.
+    /// This allows the owner to update the contract for all its users.
+    AccountId,
+}
+```
+
+Also new action is added for using previously deployed global contract:
+
+```rust
+struct UseGlobalContractAction {
+    contract_identifier: GlobalContractIdentifier,
+}
+
+enum GlobalContractIdentifier {
+    CodeHash(CryptoHash),
+    AccountId(AccountId),
+}
+```
+
+## Reference Implementation
+
+### Storage
+
+In order to have global contracts available to users on all shards we store a copy in each shard's trie.
+A new trie key is introduced for that:
+
+```rust
+pub enum TrieKey {
+    ...
+    GlobalContractCode {
+        identifier: GlobalContractCodeIdentifier,
+    },
+}
+
+pub enum GlobalContractCodeIdentifier {
+    CodeHash(CryptoHash),
+    AccountId(AccountId),
+}
+```
+
+The value is contract code bytes, similar to `TrieKey::ContractCode`.
+
+### Distribution
+
+Global contract has to be distributed to all shards after being deployed.
+This is implemented with a dedicated receipt type:
+
+```rust
+enum ReceiptEnum {
+    ...
+    GlobalContractDistribution(GlobalContractData),
+}
+
+struct GlobalContractData {
+    code: Vec<u8>,
+    id: GlobalContractIdentifier,
+}
+```
+
+`GlobalContractDistribution` receipt is generated as a result of processing `DeployGlobalContractAction`.
+Receipt distribution logic is updated to route such receipts to all shards.
+So effectively it is a part of `ShardProof` (incoming receipts) for each shard, but occurs only once in the outgoing receipts of that chunk.
+Applying `GlobalContractDistribution` receipt updates the corresponding `TrieKey::GlobalContractCode` in the trie.
```

Comment:
It would be good to test that this logic works well in case of resharding. One particular corner case would be:
* `#10` - deploy smart contract in shard X
* `#11` - shard Y and shard Z are both missing chunk
* `#12` - new shard layout after resharding where Y was split into Y1 and Y2

For X = Y, X = Z, X != Y && X != Z